Sources/CardPayments/CardClient.swift

```diff
@@ -127,11 +130,29 @@ public class CardClient: NSObject {
                     }
                 }
                 
-                let cardResult = CardResult(orderID: orderId, deepLinkURL: url)
-                self.notifySuccess(for: cardResult)
+                guard let url else {
+                    self.notifyFailure(with: CardClientError.missingDeeplinkURLError)
+                    return
+                }
+                
+                if self.getQueryStringParameter(url: url.absoluteString, param: "state") != nil
+                    && self.getQueryStringParameter(url: url.absoluteString, param: "code") != nil {
+                    let liabilityShift = self.getQueryStringParameter(url: url.absoluteString, param: "liability_shift")
+                    let cardResult = CardResult(orderID: orderId, deepLinkURL: url, liabilityShift: liabilityShift)
+                    self.notifySuccess(for: cardResult)
+                } else if self.getQueryStringParameter(url: url.absoluteString, param: "error") != nil {
+                    self.notifyFailure(with: CardClientError.threeDSVerificationError)
+                } else {
+                    self.notifyFailure(with: CardClientError.malformedDeeplinkURLError)
```

Comment:
👏 These new error cases are really helpful

Reply:
Thanks for the advice!!!